File: leetcode/0340-longest-substring-with-at-most-k-distinct-characters/solution.cpp
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        if (k == 0) {
            return 0;
        }

        unordered_map<char, int> counts;
        int left = 0, best = 0;

        for (int right = 0; right < (int)s.size(); right++) {
            counts[s[right]]++;

            while ((int)counts.size() > k) {
                char leftCh = s[left];
                if (--counts[leftCh] == 0) {
                    counts.erase(leftCh);
                }
                left++;
            }

            best = max(best, right - left + 1);
        }

        return best;
    }
};

File: leetcode/0340-longest-substring-with-at-most-k-distinct-characters/solution.py
from collections import defaultdict


class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        if k == 0:
            return 0

        counts = defaultdict(int)
        left = 0
        best = 0

        for right, ch in enumerate(s):
            counts[ch] += 1

            while len(counts) > k:
                left_ch = s[left]
                counts[left_ch] -= 1
                if counts[left_ch] == 0:
                    del counts[left_ch]
                left += 1

            best = max(best, right - left + 1)

        return best
